### ISSUE TITLE
Collector issue for known issues w/ jiti

### ISSUE THREAD

**Issue by webpro** [webpro-nl/knip#565]
There's a number of issues around using jiti. First and foremost: Jiti is awesome, Knip wouldn't nearly be as good without it!

This is a collector issue for housekeeping, and I expect those to be resolved at once. issues are documented at https://knip.dev/reference/known-issues

- https://github.com/webpro/knip/issues/346
- https://github.com/webpro/knip/issues/350
- https://github.com/webpro/knip/issues/355
- https://github.com/webpro/knip/issues/515

I think most or all of them can be potentially worked around by using bun:

```
bunx --bun knip
```

This uses the Bun runtime. Bun does not transpile, just strips types, has flexible module resolution, and also uses TS path aliases (`tsconfig.json#compilerOptions.paths`) in dynamically loaded scripts/modules.

jiti v2 might fix the issues for the Node.js runtime.

**Comment by stephenwade** [comment]
Jiti v2 has been released!

https://github.com/unjs/jiti/releases/tag/v2.0.0

**Comment by webpro** [comment]
Yay!

The jiti-v2 branch has been revived and it would be great if you could give it shot in your project(s). Install like so with your favorite package manager:

```
npm i -D https://pkg.pr.new/knip@e127e23
```

Hopefully we can remove some of the [known issues](https://knip.dev/reference/known-issues)! Planning to include this in a next patch or minor update. Let me know if you think upgrading to jiti v2 should be a major bump for Knip.

cc @stephenwade @vojtechsimetka @thenbe @EvgenyOrekhov @JoshuaKGoldberg @crystalfp @splincode

**Comment by webpro** [comment]
Also cc @deadcoder0904 @me4502 @andriyor @Codex-

**Comment by JoshuaKGoldberg** [comment]
Awesome, `knip@e127e23` works great for me on create-typescript-app. Thanks so much for this @webpro + Jiti folks!

**Comment by Codex-** [comment]
Tried it with knip reporter and it seems 👌

**Comment by stephenwade** [comment]
It works great on my project as well. I no longer need to use `bunx --bun knip`.

> Planning to include this in a next patch or minor update. Let me know if you think upgrading to jiti v2 should be a major bump for Knip.

No reason for it to be a major bump if nothing breaks in [the integration workflow](https://github.com/webpro-nl/knip/actions/workflows/integration.yml).

**Comment by me4502** [comment]
Hmm I'm testing it out on a large monorepo, and removing the `webpack: false` line from the config still fails for me. It complains with `Reason: await is only valid in async functions and the top level bodies of modules` when a top-level await is in place, but the message indicates that top level await _should_ be working.

**Comment by webpro** [comment]
> Hmm I'm testing it out on a large monorepo, and removing the `webpack: false` line from the config still fails for me when parsing a TypeScript ESM webpack config. It complains with `Reason: await is only valid in async functions and the top level bodies of modules` when a top-level await is in place, but the message indicates that top level await _should_ be working.

@me4502 Thanks for trying that out. I wrongfully assumed that `jiti` would be/include the new default ESM handling in v2 but instead that's in `jiti.import`. Should be fixed now.

```
npm i -D https://pkg.pr.new/knip@c07dfda
```

**Comment by me4502** [comment]
Thanks, that fixed the issue :) 

I did encounter some oddities around some custom webpack plugins we have, where they were compiled to JS but called out to libs that were TS- but that's probably a setup that only works for builds in the first place because ts-node doesn't care, rather than something that _should_ work. I managed to work around it by having them compile their dependencies, but just thought I'd mention it.

Otherwise this works super well, and allowed us to remove a bunch of ignored dependencies in our knip config. Thanks :)

**Comment by webpro** [comment]
:rocket: _This issue has been resolved in v5.31.0. See [Release 5.31.0](https://github.com/webpro-nl/knip/releases/tag/5.31.0) for release notes._

_Using Knip in a commercial project? Please consider [becoming a sponsor](https://knip.dev/sponsors)._

**Comment by webpro** [comment]
Thanks everyone for helping out here!
